pkg/controller/master/image/backing_image_controller.go

```diff
@@ -46,11 +65,16 @@ func (h *backingImageHandler) OnChanged(_ string, backingImage *lhv1beta2.Backin
 			toUpdate = handleFail(toUpdate, condition.Cond(harvesterv1beta1.ImageImported), fmt.Errorf(status.Message))
 			toUpdate.Status.Progress = status.Progress
 		} else if status.State == lhv1beta2.BackingImageStateReady || status.State == lhv1beta2.BackingImageStateReadyForTransfer {
-			harvesterv1beta1.ImageImported.True(toUpdate)
-			harvesterv1beta1.ImageImported.Reason(toUpdate, "Imported")
-			harvesterv1beta1.ImageImported.Message(toUpdate, status.Message)
 			toUpdate.Status.Progress = status.Progress
 			toUpdate.Status.Size = backingImage.Status.Size
+			toUpdate.Status.VirtualSize = backingImage.Status.VirtualSize
+			if toUpdate.Status.VirtualSize > 0 {
+				// We can't set ImageImported to True until we know the VirtualSize,
+				// which may happen one update after the image becomes ready.
+				harvesterv1beta1.ImageImported.True(toUpdate)
+				harvesterv1beta1.ImageImported.Reason(toUpdate, "Imported")
+				harvesterv1beta1.ImageImported.Message(toUpdate, status.Message)
+			}
```

Comment:
Just double confirm, does this mean bi could turn into a `BackingImageStateReady` state with `VirtualSize` still as zero?

That's not straightforward to me, can you help to point out the related LH source code snippet? Thanks.

Reply:
I think is [this](https://github.com/longhorn/longhorn-manager/blob/77ac26d09a1bc18e743efa2e6bb94efc7d0fa4da/controller/backing_image_controller.go#L644-L700).

Reply:
> I think is [this](https://github.com/longhorn/longhorn-manager/blob/77ac26d09a1bc18e743efa2e6bb94efc7d0fa4da/controller/backing_image_controller.go#L644-L700).

Yeah, that's it, from https://github.com/longhorn/longhorn-manager/pull/2680, which in turn relies on https://github.com/longhorn/backing-image-manager/pull/208

Unfortunately I don't have logs from when I first implemented/tested this, so I can't provide more detail right now on the comment in the code above. IIRC I observed `backingImageHandler.OnChanged()` being called twice, once when the backing image became ready, then again immediately after when the virtualSize was updated. I can re-test on Monday if that would help.

Reply:
No problem, thanks.

Reply:
I added some logging and re-tested. This block is hit twice, because the `if` condition is `status.State == lhv1beta2.BackingImageStateReady || status.State == lhv1beta2.BackingImageStateReadyForTransfer`. The first time, we come in with `status.state == BackingImageStateReadyForTransfer`, and `virtualSize == 0`. The second time, it's `status.State == lhv1beta2.BackingImageStateReady` and `virtualSize` set to the actual size of the image. I'll update the comment...

Reply:
so if we find `status.State == lhv1beta2.BackingImageStateReady`, can we confirm `backingImage.Status.VirtualSize` must not be empty?

if yes, maybe we could just skip `status.State == lhv1beta2.BackingImageStateReadyForTransfer` ?

Reply:
Yeah, good idea, that will make it a bit simpler.